pkg/oauth/oidc/interactive.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oidc
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	coreoidc "github.com/coreos/go-oidc/v3/oidc"
+	"github.com/pkg/browser"
+	"github.com/pkg/errors"
+	"github.com/segmentio/ksuid"
+	"github.com/sigstore/sigstore/pkg/oauth"
+	"golang.org/x/oauth2"
+)
+
+const oobRedirectURI = "urn:ietf:wg:oauth:2.0:oob"
+
+type browserOpener func(url string) error
+
+func doOobFlow(cfg *oauth2.Config, stateToken string, opts []oauth2.AuthCodeOption) string {
+	cfg.RedirectURL = oobRedirectURI
+	authURL := cfg.AuthCodeURL(stateToken, opts...)
+	fmt.Fprintln(os.Stderr, "Go to the following link in a browser:\n\n\t", authURL)
+	fmt.Fprintf(os.Stderr, "Enter verification code: ")
+	var code string
+	fmt.Scanln(&code)
+	return code
+}
+
+func startRedirectListener(state, htmlPage string, codeCh chan string, errCh chan error) (*http.Server, *url.URL, error) {
+	listener, err := net.Listen("tcp", "localhost:0") // ":0" == OS picks
+	if err != nil {
+		return nil, nil, err
+	}
+
+	port := listener.Addr().(*net.TCPAddr).Port
+
+	url := &url.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("localhost:%d", port),
+		Path:   "/auth/callback",
+	}
+
+	m := http.NewServeMux()
+	s := &http.Server{
+		Addr:    url.Host,
+		Handler: m,
+	}
+
+	m.HandleFunc(url.Path, func(w http.ResponseWriter, r *http.Request) {
+		// even though these are fetched from the FormValue method,
+		// these are supplied as query parameters
+		if r.FormValue("state") != state {
+			errCh <- errors.New("invalid state token")
+			return
+		}
+		codeCh <- r.FormValue("code")
+		fmt.Fprint(w, htmlPage)
+	})
+
+	go func() {
+		if err := s.Serve(listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errCh <- err
+		}
+	}()
+
+	return s, url, nil
+}
+
+func getCode(codeCh chan string, errCh chan error) (string, error) {
+	select {
+	case code := <-codeCh:
+		return code, nil
+	case err := <-errCh:
+		return "", err
+	case <-time.After(120 * time.Second):
+		return "", errors.New("timeout")
+	}
+}
+
+// newKSUID returns a globally unique, base62 (URL-safe) encoded, 27 character string.
+func newKSUID() string {
+	return ksuid.New().String()
+}
+
+func doInteractiveIDTokenFlow(ctx context.Context, cfg oauth2.Config, p *coreoidc.Provider, extraAuthCodeOpts []oauth2.AuthCodeOption, openBrowser browserOpener) (*IDToken, error) {
```

Comment:
is there a reason not to make this a method on `interactiveIDTokenSource`? all the args except for `ctx` end up being private member variables of the same struct?

Reply:
done